testing/testing.go

```diff
@@ -0,0 +1,519 @@
+		}
+
+		for j, col := range row {
+			if !assert.ObjectsAreEqualValues(e.Returns[i][j], col) {
+				return fmt.Errorf("incorrect value for expected result: row %d, column %d", i, j)
+			}
+		}
+	}
+
+	return nil
+}
+
+// Platform provides utilities and info for usage in test functions.
+// It allows users to access the database engine, get information about the
+// schema deployers, control transactions, or even directly access PostgreSQL.
+type Platform struct {
+	// Engine is the Kuneiform engine that can deploy schemas, execute actions/procedures,
+	// execute adhoc SQL, and more. It should be the primary way to interact with the database.
+	Engine common.Engine
+	// DB is the database engine that the test case is running against.
+	// It provides access directly to Postgres, and has superuser access
+	// to the underlying database. If users want to execute ad-hoc queries,
+	// they should prefer to use the Engine, which parses Kwil's SQL standard,
+	// and guarantees determinism.
+	DB sql.DB
+	// Deployer is the public identifier of the user that deployed the schemas
+	// during test setup. It can be used to execute owner-only actions and procedures.
+	// To execute owner-only actions and procedures, set the Deployer to be the
+	// *common.ExecutionData.TransactionData.Signer field when executing the
+	// action/procedure.
+	Deployer []byte
+
+	// Logger is for logging information during execution of the test.
+	Logger Logger
+
+	// lastTxid is the last transaction ID that was used.
+	lastTxid []byte
+}
+
+// Txid returns a new, unused transaction ID.
+// It is deterministic, making tests repeatable.
+func (p *Platform) Txid() string {
+	if len(p.lastTxid) == 0 {
+		b := sha256.Sum256([]byte("first txid"))
+		p.lastTxid = b[:]
+		return hex.EncodeToString(b[:])
+	}
+
+	b := sha256.Sum256(p.lastTxid)
+	p.lastTxid = b[:]
+	return hex.EncodeToString(b[:])
+}
+
+// runWithPostgres runs the callback function with a postgres container.
+func runWithPostgres(ctx context.Context, opts *Options, fn func(context.Context, *pg.DB, Logger) error) error {
+	if !opts.UseTestContainer {
+		db, err := pg.NewDB(ctx, &pg.DBConfig{
+			PoolConfig: pg.PoolConfig{
+				MaxConns: 11,
+				ConnConfig: pg.ConnConfig{
+					Host:   opts.Conn.Host,
+					Port:   opts.Conn.Port,
+					User:   opts.Conn.User,
+					Pass:   opts.Conn.Pass,
+					DBName: opts.Conn.DBName,
+				},
+			},
+		})
+		if err != nil {
+			return fmt.Errorf("error setting up database: %w", err)
+		}
+
+		defer db.Close()
+
+		return fn(ctx, db, opts.Logger)
+	}
+
+	req := testcontainers.ContainerRequest{
+		Image:        "kwildb/postgres:latest",
+		ExposedPorts: []string{"5432/tcp"},
+		Env: map[string]string{
+			"POSTGRES_HOST_AUTH_METHOD": "trust",
+		},
+		WaitingFor: wait.ForLog("database system is ready to accept connections"),
+	}
+	container, err := testcontainers.GenericContainer(ctx, testcontainers.GenericContainerRequest{
```

Comment:
@Yaiba Do you think this is the best option?  This is for kwil-cli, so in our main module.  It would be best not to have testcontainers as a dependency in our main module, but if we have to that's alright I guess.  I'm just wondering what's gained by using `testcontainers` vs. say, os.Exec to docker run docker stop etc.

Reply:
Yeah I thought about this. Another option is having a separate `kwil-test` tool that exists in the `test` package.

Reply:
yeah I agree, I would like just calling a shell command.

Reply:
If there were a complex service definition with dependencies between containers, I'd probably get the use of testcontainers for a docker compose approach, but we're just firing up the pg container.

They need docker installed regardless, so running docker commands would be preferable unless there's a notable challenge that testcontainers addresses.

If `github.com/docker/docker/client` has a reasonable dependency tree, that could work too.  https://medium.com/@Frikkylikeme/controlling-docker-with-golang-code-b213d9699998

Reply:
With just `github.com/docker/docker/client`, the go.mod deps aren't terrible:  https://gist.github.com/jchappelow/bebc12d3ed4744a93b7f25f92131ded1

Reply:
Ok I removed the test containers dependency and just made it use `exec.Command`